DRILL-6544: Timestamp value in Drill UI showed inconsistently with th…

agozhiy opened the pull request:
…e same value retrieved from sqline

Added a formatting of timestamp values instead of just calling toString().

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/ValueStringFormatter.java:
```diff
@@ -0,0 +1,41 @@
+            .appendPattern("yyyy-MM-dd HH:mm:ss")
+            .appendFraction(ChronoField.MILLI_OF_SECOND, 1, 3, true)
+            .toFormatter();
+        return value instanceof LocalDateTime ? ((LocalDateTime) value).format(formatter) : value.toString();
```
`instanceOf` check should be done prior to initializing the formatter not to waste the resources.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/ValueStringFormatter.java:
```diff
@@ -0,0 +1,41 @@
+import java.time.format.DateTimeFormatterBuilder;
+import java.time.temporal.ChronoField;
+
+public class ValueStringFormatter {
```
Maybe use more generic name, so other can add their util methods as well, ex: `WebServerUtil`.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebServerUtil.java:
```diff
@@ -0,0 +1,45 @@
+              .appendFraction(ChronoField.MILLI_OF_SECOND, 1, 3, true)
+              .toFormatter();
+          return ((LocalDateTime) value).format(formatter);
+        } else {
```
Do you need else here? Won't it go to default if there is no break?

agozhiy commented:
Addressed the review comments. Added the following options to set the display format from a pattern:
web.timestamp.display_format, 
web.date.display_format, 
web.time.display_format.
@arina-ielchiieva, could you please review?

arina-ielchiieva commented on exec/java-exec/src/test/java/org/apache/drill/exec/util/TestVectorUtil.java:
```diff
@@ -0,0 +1,154 @@
+        LocalTime.of(13, 0, 30),
+        TypeProtos.MinorType.TIME,
+        options);
+    assertTrue(formattedValue == null);
```
`assertNull`

agozhiy replied:
Fixed.

arina-ielchiieva commented on exec/java-exec/src/test/java/org/apache/drill/exec/util/TestVectorUtil.java:
```diff
@@ -0,0 +1,154 @@
+        LocalTime.of(13, 0, 30, 120000000),
+        TypeProtos.MinorType.TIME,
+        options);
+    assertTrue(formattedValue.equals("13:00:30.120"));
```
`assertEquals`

agozhiy replied:
Fixed

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/util/VectorUtil.java:
```diff
@@ -227,4 +233,69 @@ private static int getColumnWidth(int[] columnWidths, int columnIndex) {
+            try {
+              formattedValue = formatter == null ? null : ((LocalDateTime) value).format(formatter);
+            } catch (Exception e) {
+              logger.error(e.getMessage(), e);
```
You can write your own error message, including example of pattern that we could not format.

agozhiy replied:
Done.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/util/VectorUtil.java:
```diff
@@ -227,4 +233,69 @@ private static int getColumnWidth(int[] columnWidths, int columnIndex) {
+      case TIMESTAMP:
+        if (value instanceof LocalDateTime) {
+          String formatPattern = options.getString(ExecConstants.WEB_TIMESTAMP_DISPLAY_FORMAT);
+          if (!formatPattern.isEmpty()) {
```
Emptiness check can be done in `getDateTimeFormatter` value to avoid repetition.

agozhiy replied:
Moved this check to a separate method.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/util/VectorUtil.java:
```diff
@@ -227,4 +233,69 @@ private static int getColumnWidth(int[] columnWidths, int columnIndex) {
+   * @param options the OptionManager instance, not null
+   * @return the formatted value, null if failed
+   */
+  public static String formatValueVectorElement(Object value, TypeProtos.MinorType minorType, OptionManager options) {
```
Creating format instance for each record is rather expensive, plus you'll be writing errors on each incorrect record call which will pollute the log. I was thinking that  you can consider having some format class instance which will be created before you start iterating over vectors  (`for (VectorWrapper<?> vw : loader) {`). This format class will accept options in constructor and have format method where you would pass value and minor type. Formatters for timestamp / date / time will be created upon necessity and stored for further re-use.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java:
```diff
@@ -685,6 +685,18 @@ private ExecConstants() {
 
+  public static final String WEB_TIMESTAMP_DISPLAY_FORMAT = "web.timestamp.display_format";
+  public static final OptionValidator WEB_TIMESTAMP_DISPLAY_FORMAT_VALIDATOR = new StringValidator(WEB_TIMESTAMP_DISPLAY_FORMAT,
+      new OptionDescription("Display format template for timestamp. If empty then the default formatting will be used."));
```
Maybe use can mention what formatting can be applied.

agozhiy replied:
Updated the descriptions.

arina-ielchiieva commented on exec/java-exec/src/test/java/org/apache/drill/exec/util/TestValueVectorElementFormatter.java:
```diff
@@ -0,0 +1,145 @@
+public class TestValueVectorElementFormatter {
+
+  @Mock
+  OptionManager options;
```
private?

arina-ielchiieva commented on exec/java-exec/src/test/java/org/apache/drill/exec/util/TestValueVectorElementFormatter.java:
```diff
@@ -0,0 +1,145 @@
+  OptionManager options;
+
+  @Before
+  public void setup(){
```
space before `{` here and in the methods below.

agozhiy commented:
Done.

arina-ielchiieva commented:
LGTM, +.
@agozhiy thanks for making the changes.

arina-ielchiieva commented:
Merged into master with commit id 12af476247789a1decba8d4c715bd6c41d6690e8.
